### PR TITLE
Erik/logger override

### PR DESCRIPTION
**Pull request by ErikEverson**
Just some food for thought about a way to log almost all things to files as well as to Gradle

### CODE REVIEW

File: src/main/kotlin/at.phatbl.shellexec/logging/GradleLogOutputStream.kt [hunk 1]
```diff
@@ -4,11 +4,12 @@ import org.apache.commons.exec.LogOutputStream
 import org.gradle.api.GradleException
 import org.gradle.api.logging.LogLevel
 import org.gradle.api.logging.Logger
+import java.io.File
 
 /**
  * Adapter which passes log output through to Gradle's logger.
  */
-class GradleLogOutputStream(val logger: Logger, val level: LogLevel): LogOutputStream(level.ordinal) {
+class GradleLogOutputStream(val logger: Logger, val level: LogLevel,val logFiles: Array<File>?): LogOutputStream(level.ordinal) {
```
Comment (phatblat) [inline src/main/kotlin/at.phatbl.shellexec/logging/GradleLogOutputStream.kt]: Add a space after the comma. Also, add a default argument of `null` to `logFiles`, so you don't have to explicitly pass null.
```suggestion
class GradleLogOutputStream(val logger: Logger, val level: LogLevel, val logFiles: Array<File>? = null): LogOutputStream(level.ordinal) {
```

File: src/main/kotlin/at.phatbl.shellexec/ShellExec.kt [hunk 2]
```diff
@@ -24,11 +25,11 @@ open class ShellExec: DefaultTask() {
     @InputDirectory
     var workingDir: File = project.projectDir
 
-    @Internal
-    val standardOutput: OutputStream = GradleLogOutputStream(logger, LogLevel.LIFECYCLE)
+    @Input
+    var standardOutput = GradleLogOutputStream(logger, LogLevel.LIFECYCLE, null)
```
Comment (phatblat) [inline src/main/kotlin/at.phatbl.shellexec/ShellExec.kt]: The `OutputStream` type is actually more important with your change to make this property writable as the caller shouldn't need to extend `GradleLogOutputStream` in order to provide a custom `OutputStream`. `null` won't be necessary here with the default argument for `logFiles` (below).

```suggestion
    var standardOutput: OutputStream = GradleLogOutputStream(logger, LogLevel.LIFECYCLE)
```

File: src/main/kotlin/at.phatbl.shellexec/ShellExec.kt [hunk 3]
```diff
@@ -24,11 +25,11 @@ open class ShellExec: DefaultTask() {
...
 
-    @Internal
-    val errorOutput: OutputStream = GradleLogOutputStream(logger, LogLevel.ERROR)
+    @Input
+    var errorOutput = GradleLogOutputStream(logger, LogLevel.ERROR, null)
```
Comment (phatblat) [inline src/main/kotlin/at.phatbl.shellexec/ShellExec.kt]:
```suggestion
    var errorOutput: OutputStream = GradleLogOutputStream(logger, LogLevel.ERROR)
```

**Review by phatblat — CHANGES_REQUESTED** [review]
This is a great change to add "tee"-like functionality! I just have some syntax suggestions.

**Comment by ErikEverson** [comment]
A note about adding back in the outputStream as the required type. When I tested this with a composite project I get the following compile errors.

Screen Shot 2019-11-06 at 8 01 47 AM

**Comment by ErikEverson** [comment]
@phatblat PR is ready to be looked at. I'm not sure if the Gradle log stuff will work once the project is a JAR but I think we should try that before changing away from Output stream

**Review by phatblat — APPROVED** [review]

**Pull request merged**